crates/topos-p2p/src/behaviour/transmission/codec.rs

```diff
@@ -1,57 +1,107 @@
     type Request = TransmissionRequest;
-    type Response = TransmissionResponse;
+    type Response = Result<TransmissionResponse, ()>;
 
     async fn read_request<T>(
         &mut self,
-        _: &Self::Protocol,
-        io: &mut T,
+        protocol: &Self::Protocol,
+        mut io: &mut T,
     ) -> std::io::Result<Self::Request>
     where
         T: AsyncRead + Unpin + Send,
     {
-        let vec = read_length_prefixed(io, 1_000_000_000).await?;
-        Ok(TransmissionRequest(vec))
+        // Read the length.
+        let length = unsigned_varint::aio::read_usize(&mut io)
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
+        if length > usize::try_from(MAX_PACKET_SIZE).unwrap_or(usize::MAX) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!(
+                    "Request size exceeds limit: {} > {}",
+                    length, MAX_PACKET_SIZE
+                ),
+            ));
+        }
+
+        // Read the payload.
+        let mut buffer = vec![0; length];
+        io.read_exact(&mut buffer).await?;
+
+        match protocol.as_ref() {
+            SYNCHRONIZER_PROTOCOL => Ok(TransmissionRequest::Synchronizer(buffer)),
+            _ => Ok(TransmissionRequest::Transmission(buffer)),
+        }
     }
 
     async fn read_response<T>(
         &mut self,
-        _: &Self::Protocol,
-        io: &mut T,
+        protocol: &Self::Protocol,
+        mut io: &mut T,
     ) -> std::io::Result<Self::Response>
     where
         T: AsyncRead + Unpin + Send,
     {
-        let vec = read_length_prefixed(io, 1_000_000).await?;
+        // Read the length of the payload.
+        let length = match unsigned_varint::aio::read_usize(&mut io).await {
+            Ok(l) => l,
+            Err(unsigned_varint::io::ReadError::Io(err))
+                if matches!(err.kind(), io::ErrorKind::UnexpectedEof) =>
+            {
+                return Ok(Err(()))
+            }
+            Err(err) => return Err(io::Error::new(io::ErrorKind::InvalidInput, err)),
+        };
 
-        if vec.is_empty() {
-            return Err(io::ErrorKind::UnexpectedEof.into());
+        if length > usize::try_from(MAX_PACKET_SIZE).unwrap_or(usize::MAX) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!(
+                    "Response size exceeds limit: {} > {}",
+                    length, MAX_PACKET_SIZE
+                ),
+            ));
         }
 
-        Ok(TransmissionResponse(vec))
+        // Read the payload itself.
+        let mut buffer = vec![0; length];
+        io.read_exact(&mut buffer).await?;
+
+        let res = match protocol.as_ref() {
+            SYNCHRONIZER_PROTOCOL => Ok(TransmissionResponse::Synchronizer(buffer)),
+            _ => Ok(TransmissionResponse::Transmission(buffer)),
+        };
+
+        Ok(res)
```

Comment:
Yes it is. It allows us to ACK network transmission but respond with failure if needed.

Reply:
So if the `Result` is `Ok(Err)` it will be sent but just `Err` would just be logged and not replied back? What is stopping us from sending back an error?

Reply:
The `Result<_, _>` is the codec result, the `Result< Result< _, _> >` is the underlying type response. If the request is invalid or the sender isn't authorized we don't respond and close the channel. Dealing with proper `Response errors` are handled directly at the message level.